### PR TITLE
fix ReadTimeoutException: null

### PR DESCRIPTION
**Pull request by naah69**
// Describe your PR here; eg. Fixes #issueNo
#1222 
Make sure that:

- [x] You have read the [contribution guidelines](https://dromara.org/projects/soul/contributor/).
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] Your local test passed `mvn clean install -Dmaven.javadoc.skip=true`.

### CODE REVIEW

File: soul-spring-boot-starter/soul-spring-boot-starter-plugin/soul-spring-boot-starter-plugin-httpclient/src/main/java/org/dromara/soul/springboot/starter/plugin/httpclient/HttpClientPluginConfiguration.java [hunk 1]
```diff
@@ -110,17 +107,11 @@ public HttpClient httpClient(final HttpClientProperties properties) {
...
-                        Map<String, ChannelHandler> nameChannelHandlerMap = pipeline.toMap();
-                        // The write and read timeouts are serving as generic socket idle state handlers.
-                        if (!nameChannelHandlerMap.containsKey("write_timeout")) {
-                            pipeline.addFirst("write_timeout", new WriteTimeoutHandler(properties.getWriteTimeout(), TimeUnit.MILLISECONDS));
-                        }
-                        if (!nameChannelHandlerMap.containsKey("read_timeout")) {
-                            pipeline.addFirst("read_timeout", new ReadTimeoutHandler(properties.getReadTimeout(), TimeUnit.MILLISECONDS));
-                        }
+                        c.addHandlerLast(new WriteTimeoutHandler(properties.getWriteTimeout(), TimeUnit.MILLISECONDS));
+                        c.addHandlerLast(new ReadTimeoutHandler(properties.getReadTimeout(), TimeUnit.MILLISECONDS));
```
Comment (dengliming) [inline soul-spring-boot-starter/soul-spring-boot-starter-plugin/soul-spring-boot-starter-plugin-httpclient/src/main/java/org/dromara/soul/springboot/starter/plugin/httpclient/HttpClientPluginConfiguration.java:114]: Please make sure that [this exception](https://github.com/dromara/soul/issues/1209)  will not occur.

File: soul-spring-boot-starter/soul-spring-boot-starter-plugin/soul-spring-boot-starter-plugin-httpclient/src/main/java/org/dromara/soul/springboot/starter/plugin/httpclient/HttpClientPluginConfiguration.java [hunk 2]
```diff
@@ -110,17 +107,11 @@ public HttpClient httpClient(final HttpClientProperties properties) {
                                     .to(builder::nonProxyHosts);
                         });
                     }
-                    tcpClient = tcpClient.option(ChannelOption.SO_TIMEOUT, 5000);
+                    tcpClient = tcpClient.option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 5000);
```
Comment (dengliming) [inline soul-spring-boot-starter/soul-spring-boot-starter-plugin/soul-spring-boot-starter-plugin-httpclient/src/main/java/org/dromara/soul/springboot/starter/plugin/httpclient/HttpClientPluginConfiguration.java:110]: It's already set up [here](https://github.com/dromara/soul/blob/000e54d8169c4db70f6c18048ce67eac07cd4ebf/soul-spring-boot-starter/soul-spring-boot-starter-plugin/soul-spring-boot-starter-plugin-httpclient/src/main/java/org/dromara/soul/springboot/starter/plugin/httpclient/HttpClientPluginConfiguration.java#L90-L95)

**Review by dengliming — CHANGES_REQUESTED** [review]

File: soul-spring-boot-starter/soul-spring-boot-starter-plugin/soul-spring-boot-starter-plugin-httpclient/src/main/java/org/dromara/soul/springboot/starter/plugin/httpclient/HttpClientPluginConfiguration.java (hunk 1 above)
Comment (naah69) [reply soul-spring-boot-starter/soul-spring-boot-starter-plugin/soul-spring-boot-starter-plugin-httpclient/src/main/java/org/dromara/soul/springboot/starter/plugin/httpclient/HttpClientPluginConfiguration.java:113]: yes, i had tested it.

Comment (dengliming) [reply soul-spring-boot-starter/soul-spring-boot-starter-plugin/soul-spring-boot-starter-plugin-httpclient/src/main/java/org/dromara/soul/springboot/starter/plugin/httpclient/HttpClientPluginConfiguration.java:113]: @naah69 What does this PR actually fix? It's `read_timeout/write_timeout` that doesn't work? BTW, what is the actual effect of `addHandlerLast ` and `addHandlerFirst`?

Comment (naah69) [reply soul-spring-boot-starter/soul-spring-boot-starter-plugin/soul-spring-boot-starter-plugin-httpclient/src/main/java/org/dromara/soul/springboot/starter/plugin/httpclient/HttpClientPluginConfiguration.java:113]: yes,read_timeout/write_timeout doesn't work.it allways make error that is `io.netty.handler.timeout.ReadTimeoutException: null`,then the request will be 500.

from Reactor Netty 0.8 and Spring Framework 5.1,the handler will be set by below code.i fix it to make code work.

```
TcpClient tcpClient = TcpClient.create()
                 .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 1000)
                 .doOnConnected(connection ->
                         connection.addHandlerLast(new ReadTimeoutHandler(10))
                                   .addHandlerLast(new WriteTimeoutHandler(10)));
WebClient webClient = WebClient.builder()
    .clientConnector(new ReactorClientHttpConnector(HttpClient.from(tcpClient)))
    .build();
```

Comment (dengliming) [reply soul-spring-boot-starter/soul-spring-boot-starter-plugin/soul-spring-boot-starter-plugin-httpclient/src/main/java/org/dromara/soul/springboot/starter/plugin/httpclient/HttpClientPluginConfiguration.java:113]: So, what is the returned data when timeout occurs?

Comment (dengliming) [reply soul-spring-boot-starter/soul-spring-boot-starter-plugin/soul-spring-boot-starter-plugin-httpclient/src/main/java/org/dromara/soul/springboot/starter/plugin/httpclient/HttpClientPluginConfiguration.java:113]: @naah69 Any progress on this?

**Comment by yu199195** [comment]
please checked ci

**Review by dengliming — APPROVED** [review]
Thanks!

**Comment by linchuyao** [comment]
There is still a problem
EE01E76E-02A6-4D29-9377-D30DAB3DF0E6

**Comment by dengliming** [comment]
@linchuyao That is not a `NullPointException`.

**Comment by linchuyao** [comment]
why now timeOutException soul will response 500 errorCode

**Comment by dengliming** [comment]
No particular reason. If you have a better suggestion, you can submit an issue to discuss this case. : )

**Comment by naah69** [comment]
> There is still a problem
> EE01E76E-02A6-4D29-9377-D30DAB3DF0E6
> 

this pr exactly fix it

**Pull request merged**